### ISSUE TITLE
unhappy exception on Connection Refused to AMQP at Shared Services Canada

### ISSUE THREAD

**Issue by idallen** [JohnTheNerd/PySarra#7]
I got exception tracebacks and a program exit when the program got a Connection Refused from AMQP.
Could the program please quietly catch the error, sleep, and retry a few times before giving up?
~~~
2022-11-09 19:30:38,554 - ERROR - pika.adapters.utils.io_services_utils - _consume: protocol.data_received() failed: error=URLError(ConnectionRefusedError(111, 'Connection refused')); <ssl.SSLSocket fd=9, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('192.168.9.250', 56436), raddr=('205.189.10.47', 5671)> 
Traceback (most recent call last):
  File "/usr/lib/python3.8/urllib/request.py", line 1354, in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
  File "/usr/lib/python3.8/http/client.py", line 1256, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1302, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1251, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.8/http/client.py", line 1011, in _send_output
    self.send(msg)
  File "/usr/lib/python3.8/http/client.py", line 951, in send
    self.connect()
  File "/usr/lib/python3.8/http/client.py", line 1418, in connect
    super().connect()
  File "/usr/lib/python3.8/http/client.py", line 922, in connect
    self.sock = self._create_connection(
  File "/usr/lib/python3.8/socket.py", line 808, in create_connection
    raise err
  File "/usr/lib/python3.8/socket.py", line 796, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

[... dozens more traceback lines...]

2022-11-09 19:30:38,591 - INFO - pika.adapters.utils.io_services_utils - _initiate_abort: _AsyncTransportBase._initate_abort(): Initiating abrupt asynchronous transport shutdown: state=1; error=URLError(ConnectionRefusedError(111, 'Connection refused')); <ssl.SSLSocket fd=9, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('192.168.9.250', 56436), raddr=('205.189.10.47', 5671)>
2022-11-09 19:30:38,591 - INFO - pika.adapters.utils.io_services_utils - _deactivate: Deactivating transport: state=1; <ssl.SSLSocket fd=9, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('192.168.9.250', 56436), raddr=('205.189.10.47', 5671)>
2022-11-09 19:30:38,591 - ERROR - pika.adapters.base_connection - _proto_connection_lost: connection_lost: StreamLostError: ("Stream connection lost: URLError(ConnectionRefusedError(111, 'Connection refused'))",)
2022-11-09 19:30:38,591 - INFO - pika.connection - _on_stream_terminated: AMQP stack terminated, failed to connect, or aborted: opened=True, error-arg=StreamLostError: ("Stream connection lost: URLError(ConnectionRefusedError(111, 'Connection refused'))",); pending-error=None
2022-11-09 19:30:38,591 - ERROR - pika.callback - process: Calling <bound method ECConsumer.on_channel_closed of <__main__.ECConsumer object at 0x7fbaaf142b50>> for "1:_on_channel_close" failed
Traceback (most recent call last):
  File "/home/idallen/.local/lib/python3.8/site-packages/pika/callback.py", line 233, in process
    callback(*args, **keywords)
TypeError: on_channel_closed() missing 1 required positional argument: 'reply_text'
2022-11-09 19:30:38,591 - ERROR - pika.adapters.utils.io_services_utils - _connection_lost_notify_async: protocol.connection_lost(URLError(ConnectionRefusedError(111, 'Connection refused'))) failed: exc=TypeError("on_channel_closed() missing 1 required positional argument: 'reply_text'"); <ssl.SSLSocket fd=9, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('192.168.9.250', 56436), raddr=('205.189.10.47', 5671)>

[... dozens more traceback lines...]

2022-11-09 19:30:38,592 - INFO - pika.adapters.utils.io_services_utils - _close_and_finalize: Closing transport socket and unlinking: state=2; <ssl.SSLSocket fd=9, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('192.168.9.250', 56436), raddr=('205.189.10.47', 5671)>
2022-11-09 19:30:38,592 - ERROR - pika.adapters.utils.io_services_utils - log_exception_func_wrap: Wrapped func exited with exception. Caller's stack:
Traceback (most recent call last):
  File "/home/idallen/.local/lib/python3.8/site-packages/pika/diagnostic_utils.py", line 53, in log_exception_func_wrap
    return func(*args, **kwargs)
  File "/home/idallen/.local/lib/python3.8/site-packages/pika/adapters/utils/io_services_utils.py", line 972, in _connection_lost_notify_async
    self._protocol.connection_lost(error)
  File "/home/idallen/.local/lib/python3.8/site-packages/pika/adapters/base_connection.py", line 431, in _proto_connection_lost
    self._on_stream_terminated(error)
  File "/home/idallen/.local/lib/python3.8/site-packages/pika/connection.py", line 2080, in _on_stream_terminated
    self._channels[channel]._on_close_meta(self._error)
  File "/home/idallen/.local/lib/python3.8/site-packages/pika/channel.py", line 1124, in _on_close_meta
    self._transition_to_closed()
  File "/home/idallen/.local/lib/python3.8/site-packages/pika/channel.py", line 1068, in _transition_to_closed
    self.callbacks.process(self.channel_number, '_on_channel_close',
  File "/home/idallen/.local/lib/python3.8/site-packages/pika/callback.py", line 60, in wrapper
    return function(*tuple(args), **kwargs)
  File "/home/idallen/.local/lib/python3.8/site-packages/pika/callback.py", line 92, in wrapper
    return function(*args, **kwargs)
  File "/home/idallen/.local/lib/python3.8/site-packages/pika/callback.py", line 233, in process
    callback(*args, **keywords)
TypeError: on_channel_closed() missing 1 required positional argument: 'reply_text'
Traceback (most recent call last):

[... dozens more traceback lines...]
~~~

**Comment by JohnTheNerd** [comment]
I don't believe this should be PySarra's responsibility, as there are several other applications that are much more suitable to restarting services as needed. May I recommend running it in systemd (or init.d if your system does not support it)?

**Comment by idallen** [comment]
I just run it in a loop in a shell script.  Since the error is expected from time-to-time, could PySarra handle it quietly without the need for a traceback?  The traceback is not useful in this case.

**Comment by JohnTheNerd** [comment]
mostly resolved by #10, probably shouldn't add more complexity just to avoid exceptions passing between threads
